src/components/abstractComponent.ts

```diff
@@ -0,0 +1,579 @@
+   * is not guaranteed to be reflected by the DOM at the time the callback is fired.
+   *
+   * @param {IResizeHandler} [resizeHandler] Callback to be called when component resizes
+   */
+  onResize(resizeHandler: IResizeHandler): this;
+  /**
+   * Renders the Component without waiting for the next frame. This method is a no-op on
+   * Component, Table, and Group; render them immediately with .renderTo() instead.
+   *
+   * @returns {IComponent} The calling Component.
+   */
+  renderImmediately(): this;
+  /**
+   * Causes the Component to re-layout and render.
+   *
+   * This function should be called when a CSS change has occured that could
+   * influence the layout of the Component, such as changing the font size.
+   *
+   * @returns {IComponent} The calling Component.
+   */
+  redraw(): this;
+  /**
+   * Renders the Component to a given D
+   *
+   * @param {D} container The object to render to
+   * @returns {IComponent} The calling Component.
+   */
+  renderTo(container: D): this;
+  /**
+   * Gets the x alignment of the Component.
+   */
+  xAlignment(): string;
+  /**
+   * Sets the x alignment of the Component.
+   *
+   * @param {string} xAlignment The x alignment of the Component ("left"/"center"/"right").
+   * @returns {IComponent} The calling Component.
+   */
+  xAlignment(xAlignment: string): this;
+  xAlignment(xAlignment?: string): any;
+  /**
+   * Gets the y alignment of the Component.
+   */
+  yAlignment(): string;
+  /**
+   * Sets the y alignment of the Component.
+   *
+   * @param {string} yAlignment The y alignment of the Component ("top"/"center"/"bottom").
+   * @returns {IComponent} The calling Component.
+   */
+  yAlignment(yAlignment: string): this;
+  yAlignment(yAlignment?: string): any;
+  /**
+   * Checks if the Component has a given CSS class.
+   *
+   * @param {string} cssClass The CSS class to check for.
+   */
+  hasClass(cssClass: string): boolean;
+  /**
+   * Adds a given CSS class to the Component.
+   *
+   * @param {string} cssClass The CSS class to add.
+   * @returns {IComponent} The calling Component.
+   */
+  addClass(cssClass: string): this;
+  /**
+   * Removes a given CSS class from the Component.
+   *
+   * @param {string} cssClass The CSS class to remove.
+   * @returns {IComponent} The calling Component.
+   */
+  removeClass(cssClass: string): this;
+  /**
+   * Checks if the Component has a fixed width or if it grows to fill available space.
+   * Returns false by default on the base Component class.
+   */
+  fixedWidth(): boolean;
+  /**
+   * Checks if the Component has a fixed height or if it grows to fill available space.
+   * Returns false by default on the base Component class.
+   */
+  fixedHeight(): boolean;
+  /**
+   * Detaches a Component from the DOM. The Component can be reused.
+   *
+   * This should only be used if you plan on reusing the calling Component. Otherwise, use destroy().
+   *
+   * @returns The calling Component.
+   */
+  detach(): this;
```

Comment:
Yea I'm not sure ... I think its for moving components around the hierarchy (without destroying them) .. but I don't know when it happens or why we need it.